torch/optim/optimizer.py

```diff
@@ -383,9 +384,11 @@ def zero_grad(self, set_to_none: bool = False):
                             else:
                                 p.grad.requires_grad_(False)
                             if (not foreach or p.grad.is_sparse):
-                                p.grad.zero_()
+                                tmp_group.append(p)
```

Comment:
The logic to use the foreach implementation is already there just below.
Just change the condition above to be `if p.grad.is_sparse:`
The problem remains the same as before though: not all backend implement these foreach ops and so you should only use them when possible: meaning on CUDA and for the proper dtypes.

Reply:
Since my internship ends, I automatically lost permission to push new commits to this PR. So have to add comments in the code review mode.

Reply:
Ho that is unfortunate! You can push the change on a branch in your fork of pytorch on github to be able to send code!